<a href="https://colab.research.google.com/github/SitiMutmainah15/PEMBELAJARAN-MESIN_GENAP_2026/blob/main/JS05/Tugas_Lab2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import pandas as pd

# Load data
df = pd.read_csv('spam.csv', encoding='latin-1')

# Ambil hanya kolom v1 dan v2
df = df.iloc[:, :2]

# Rename kolom
df = df.rename(columns={
    'v1': 'Labels',
    'v2': 'SMS'
})

# Encoding label
df['Labels'] = df['Labels'].map({
    'spam': 1,
    'ham': 0
})

df.head()

,Labels,SMS
0,0,"Go until jurong point, crazy.. Available only ..."
1,0,Ok lar... Joking wif u oni...
2,1,Free entry in 2 a wkly comp to win FA Cup fina...
3,0,U dun say so early hor... U c already then say...
4,0,"Nah I don't think he goes to usf, he lives aro..."


In [6]:
X = df['SMS'].values
y = df['Labels'].values

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=50
)

In [7]:
from sklearn.feature_extraction.text import CountVectorizer

# CountVectorizer dengan stop words
bow = CountVectorizer(stop_words='english')

# Fit hanya pada data training
X_train_bow = bow.fit_transform(X_train)

# Transform data testing
X_test_bow = bow.transform(X_test)

print("Jumlah fitur:", len(bow.get_feature_names_out()))
print("Dimensi data training:", X_train_bow.shape)

Jumlah fitur: 7466
Dimensi data training: (4457, 7466)


In [8]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

mnb_bow = MultinomialNB()

# Training
mnb_bow.fit(X_train_bow, y_train)

# Prediksi
y_pred_bow = mnb_bow.predict(X_test_bow)

# Evaluasi
acc_bow = accuracy_score(y_test, y_pred_bow)

print("Akurasi CountVectorizer:", acc_bow)
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_bow))
print("\nClassification Report:")
print(classification_report(y_test, y_pred_bow))

Akurasi CountVectorizer: 0.9829596412556054

Confusion Matrix:
[[951   3]
 [ 16 145]]

Classification Report:
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       954
           1       0.98      0.90      0.94       161

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115



CountVectorizer menghasilkan akurasi 98,30% dan recall spam 90%, sehingga mampu mengklasifikasikan spam dan ham dengan sangat baik.

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer

# TF-IDF dengan stop words
tfidf = TfidfVectorizer(stop_words='english')

# Fit dan transform data training
X_train_tfidf = tfidf.fit_transform(X_train)

# Transform data testing
X_test_tfidf = tfidf.transform(X_test)

print("Jumlah fitur:", len(tfidf.get_feature_names_out()))
print("Dimensi data training:", X_train_tfidf.shape)

Jumlah fitur: 7466
Dimensi data training: (4457, 7466)


In [10]:
# Model Multinomial Naive Bayes
mnb_tfidf = MultinomialNB()

# Training
mnb_tfidf.fit(X_train_tfidf, y_train)

# Prediksi
y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)

# Evaluasi
acc_tfidf = accuracy_score(y_test, y_pred_tfidf)

print("Akurasi TF-IDF:", acc_tfidf)
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_tfidf))
print("\nClassification Report:")
print(classification_report(y_test, y_pred_tfidf))

Akurasi TF-IDF: 0.9605381165919282

Confusion Matrix:
[[954   0]
 [ 44 117]]

Classification Report:
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       954
           1       1.00      0.73      0.84       161

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115



TF-IDF menghasilkan akurasi 96,05% dengan recall spam 73%, sehingga masih terdapat lebih banyak spam yang salah diklasifikasikan sebagai ham.

Pada percobaan ini, CountVectorizer merupakan fitur yang lebih optimal untuk klasifikasi spam.csv menggunakan Multinomial Naive Bayes.